### DISCUSSION TITLE
An Integrated Keyboard

### DISCUSSION THREAD

**Discussion by ghost** [termux/termux-app#2232] [Ideas]
Hi there Dev Team,

Since Hackers Keyboard for android has become an ancient project running on sdk api level 9 its time for termux to think for a new alternative and there's none. Why not build an integrated Keyboard like any IDE app does for coding. I think this will be a great idea trust me gboard, swiftkey, samsung keyboard none work better in termux and we need a Keyboard to operate in termux right. Pls consider this.

Regards.

**Comment by agnostic-apollo** [comment]
Those keyboards work fine, some devices may need [`enforce-char-based-input=true`](https://wiki.termux.com/wiki/Terminal_Settings) but they work. Implementing or even porting a new keyboard when there is no added benefit is likely something no maintainer would do. Also those keyboards have multi-language support as well. If hackers keyboard is old, then someone should fork it and update, although it likely still works too. Termux already provides [extra keys](https://wiki.termux.com/wiki/Touch_Keyboard) which are sufficient to supplement a standard keyboard that doesn't have special keys.

**Reply by creepy-pasta101** [reply]
Yes.... First rule of developers.. Don't try to fix something that has already been fixed..

**Reply by agnostic-apollo** [reply]
pretty much...

**Reply by SDRausty** [reply]
>Don't try to fix something that has already been fixed..

What if it is fixed, and it is still broken?  If you do agree that the Android default keyboards are broken, can any of these Android keyboards https://github.com/BuildAPKs/buildAPKs.keyboards be integrated as integrated keyboard addons with Termux?

**Reply by SDRausty** [reply]
> the Android default keyboards are broken as I have stated

It seems that they are broken in a way like a moth and a candle flame in darkened room without sunshine, nor twilight about candle flame in darkened room of dancing in air, circling, fluttering with wings sizzling moths.

**Reply by agnostic-apollo** [reply]
Why would termux ship random third party keyboards? Users can install whatever keyboard they want from relevant stores or sites.

**Reply by SDRausty** [reply]
>Why would termux ship random third party keyboards? 

These keyboards are not random, and they do build in Termux at present.  The open source code could be one more Termux addon, and with time possibly a few;  Integrated Keyboard user requests is also leads the answer to why, instead of why not.

>Users can install whatever keyboard they want from relevant stores or sites.

This is true, and makes this topic mute as we can also build virtual open source Android keyboards in Termux in our devices too;  Thank you!

**Comment by Elidevin** [comment]
![integrated_keyboard](https://user-images.githubusercontent.com/47048455/131264458-29b5df79-09c6-4e76-97c3-5c68c3fe8963.png)

`~/.termux/termux.properties`
```javascript
extra-keys = [ \
	[ \
		ESC, \
		{ key: '`', popup: '~', display: "` ~" }, \
		TAB, \
		{ key: F1, popup:F7 }, \
		{ key: F2, popup:F8 }, \
		{ key: F3, popup:F9 }, \
		{ key: F4, popup:F10 }, \
		{ key: F5, popup:F11 }, \
		{ key: F6, popup:F12 }, \
		'', \
		DELETE, \
		BACKSPACE \
	], \
	[ \
		{ key: '1', popup: '!', display: "1 !" }, \
		{ key: '2', popup: '@', display: "2 @" }, \
		{ key: '3', popup: '#', display: "3 #" }, \
		{ key: '4', popup: '$', display: "4 $" }, \
		{ key: '5', popup: '%', display: "5 %" }, \
		{ key: '6', popup: '^', display: "6 ^" }, \
		{ key: '7', popup: '&', display: "7 &" }, \
		{ key: '8', popup: '*', display: "8 *" }, \
		{ key: '9', popup: '(', display: "9 (" }, \
		{ key: '0', popup: ')', display: "0 )" }, \
		{ key: '-', popup: '_', display: "- _" }, \
		{ key: '=', popup: '+', display: "= +" } \
	], \
	[ \
		{ key: 'q', popup: 'Q'}, \
		{ key: 'w', popup: 'W'}, \
		{ key: 'e', popup: 'E'}, \
		{ key: 'r', popup: 'R'}, \
		{ key: 't', popup: 'T'}, \
		{ key: 'y', popup: 'Y'}, \
		{ key: 'u', popup: 'U'}, \
		{ key: 'i', popup: 'I'}, \
		{ key: 'o', popup: 'O'}, \
		{ key: 'p', popup: 'P'}, \
		{ key: '\[', popup: '\{', display: "\[ \{" }, \
		{ key: '\]', popup: '\}', display: "\] \}" } \
	], \
	[ \
		{ key: 'a', popup: 'A'}, \
		{ key: 's', popup: 'S'}, \
		{ key: 'd', popup: 'D'}, \
		{ key: 'f', popup: 'F'}, \
		{ key: 'g', popup: 'G'}, \
		{ key: 'h', popup: 'H'}, \
		{ key: 'j', popup: 'J'}, \
		{ key: 'k', popup: 'K'}, \
		{ key: 'l', popup: 'L'}, \
		{ key: ';', popup: ':', display: "; :" }, \
		{ key: "\'", popup: '\"', display:"\' \'\'" }, \
		{ key: ENTER, popup: '\n' } \
	], \
	[ \
		{ key: 'z', popup: 'Z'}, \
		{ key: 'x', popup: 'X'}, \
		{ key: 'c', popup: 'C'}, \
		{ key: 'v', popup: 'V'}, \
		{ key: 'b', popup: 'B'}, \
		{ key: 'n', popup: 'N'}, \
		{ key: 'm', popup: 'M'}, \
		{ key: ',', popup: '<', display: ", <" }, \
		{ key: '.', popup: '>', display: ". >" }, \
		{ key: '/', popup: '?', display: "/ ?" }, \
		{ key: UP, popup: PGUP }, \
		{ key: '|', popup: "\\\\", display: "| \\\\" } \
	], \
	[ \
		CTRL, \
		ALT, \
		KEYBOARD, \
		' ', \
		' ', \
		' ', \
		' ', \
		' ', \
		' ', \
		{ key: LEFT, popup: HOME }, \
		{ key: DOWN, popup: PGDN }, \
		{ key: RIGHT, popup: END } \
	] \
]
```

**Reply by SDRausty** [reply]
Thank you very much for sharing a resolution @Elidevin, and thank you for making this resolution possible open source Termux!  Can this information about our integrated keyboard be added somewhere to our help pages here https://wiki.termux.com/index.php?search=keyboard please?

**Reply by Elidevin** [reply]
yeap :)

**Reply by SDRausty** [reply]
👍 🎵 🎶

**Reply by Yonle** [reply]
What in the world did you just made. Also amazing job

**Reply by ghost** [reply]
Man popup characters are appearing after swiping each characters upward is it so or am I doing something wrong although no popup characters are showing while swiping up.

Also can u tell me a way to disable the inbuilt virtual keyboard of android on termux.

**Reply by Elidevin** [reply]
https://user-images.githubusercontent.com/47048455/131310809-58d2a67e-1e2c-4ebe-ab88-21791abe0823.mp4

**Reply by ghost** [reply]
Don't answer for the second part I already hide the default soft keyboard from 

long press on termux screen > more > settings > termux > Terminal I/O > Enable Softkeyboard < this one disabled 

Now I only have ur beautiful keyboard also the virtual keyboard is not getting popped up frequently actually not at all.

Thanks Man U are life saver.

**Reply by ghost** [reply]
> 
> https://user-images.githubusercontent.com/47048455/131310809-58d2a67e-1e2c-4ebe-ab88-21791abe0823.mp4
> 
> 

Thanks Man got it

**Reply by ghost** [reply]
Is there anyway to display the characters in lowercase instead of uppercase I tried 
{ key: "q", popup: "Q", display: "q" } /,

but q nor showing in lowercase.

Thanks

**Reply by agnostic-apollo** [reply]
Android `Button` `View` uppercases all text by default. Hence, currently termux displays it as uppercase only. An option can be added to disable automatic uppercasing if someone wants that.

**Reply by ghost** [reply]
> Android `Button` `View` uppercases all text by default. Hence, currently termux displays it as uppercase only. An option can be added to disable automatic uppercasing if someone wants that.

what option man can I be able to add it?

**Reply by agnostic-apollo** [reply]
Currently no option exists. Can be added for next release.

**Reply by SDRausty** [reply]
>Currently no option exists. Can be added for next release.

Please reference ["Add overlay keyboard with Alt, Ctrl, Tab, arrow keys (and a few more keys) #97"](https://github.com/termux/termux-app/issues/12), and thank you for considering adding to the next release 🎹

**Reply by Archenoth** [reply]
Oh heck, this is amazing! Thank you for sharing, @Elidevin..!

**Comment by ghost** [comment]
> ![integrated_keyboard](https://user-images.githubusercontent.com/47048455/131264458-29b5df79-09c6-4e76-97c3-5c68c3fe8963.png)
> 
> `~/.termux/termux.properties`
> ```javascript
> extra-keys = [ \
> 	[ \
> 		ESC, \
> 		{ key: '`', popup: '~', display: "` ~" }, \
> 		TAB, \
> 		{ key: F1, popup:F7 }, \
> 		{ key: F2, popup:F8 }, \
> 		{ key: F3, popup:F9 }, \
> 		{ key: F4, popup:F10 }, \
> 		{ key: F5, popup:F11 }, \
> 		{ key: F6, popup:F12 }, \
> 		'', \
> 		DELETE, \
> 		BACKSPACE \
> 	], \
> 	[ \
> 		{ key: '1', popup: '!', display: "1 !" }, \
> 		{ key: '2', popup: '@', display: "2 @" }, \
> 		{ key: '3', popup: '#', display: "3 #" }, \
> 		{ key: '4', popup: '$', display: "4 $" }, \
> 		{ key: '5', popup: '%', display: "5 %" }, \
> 		{ key: '6', popup: '^', display: "6 ^" }, \
> 		{ key: '7', popup: '&', display: "7 &" }, \
> 		{ key: '8', popup: '*', display: "8 *" }, \
> 		{ key: '9', popup: '(', display: "9 (" }, \
> 		{ key: '0', popup: ')', display: "0 )" }, \
> 		{ key: '-', popup: '_', display: "- _" }, \
> 		{ key: '=', popup: '+', display: "= +" } \
> 	], \
> 	[ \
> 		{ key: 'q', popup: 'Q'}, \
> 		{ key: 'w', popup: 'W'}, \
> 		{ key: 'e', popup: 'E'}, \
> 		{ key: 'r', popup: 'R'}, \
> 		{ key: 't', popup: 'T'}, \
> 		{ key: 'y', popup: 'Y'}, \
> 		{ key: 'u', popup: 'U'}, \
> 		{ key: 'i', popup: 'I'}, \
> 		{ key: 'o', popup: 'O'}, \
> 		{ key: 'p', popup: 'P'}, \
> 		{ key: '\[', popup: '\{', display: "\[ \{" }, \
> 		{ key: '\]', popup: '\}', display: "\] \}" } \
> 	], \
> 	[ \
> 		{ key: 'a', popup: 'A'}, \
> 		{ key: 's', popup: 'S'}, \
> 		{ key: 'd', popup: 'D'}, \
> 		{ key: 'f', popup: 'F'}, \
> 		{ key: 'g', popup: 'G'}, \
> 		{ key: 'h', popup: 'H'}, \
> 		{ key: 'j', popup: 'J'}, \
> 		{ key: 'k', popup: 'K'}, \
> 		{ key: 'l', popup: 'L'}, \
> 		{ key: ';', popup: ':', display: "; :" }, \
> 		{ key: "\'", popup: '\"', display:"\' \'\'" }, \
> 		{ key: ENTER, popup: '\n' } \
> 	], \
> 	[ \
> 		{ key: 'z', popup: 'Z'}, \
> 		{ key: 'x', popup: 'X'}, \
> 		{ key: 'c', popup: 'C'}, \
> 		{ key: 'v', popup: 'V'}, \
> 		{ key: 'b', popup: 'B'}, \
> 		{ key: 'n', popup: 'N'}, \
> 		{ key: 'm', popup: 'M'}, \
> 		{ key: ',', popup: '<', display: ", <" }, \
> 		{ key: '.', popup: '>', display: ". >" }, \
> 		{ key: '/', popup: '?', display: "/ ?" }, \
> 		{ key: UP, popup: PGUP }, \
> 		{ key: '|', popup: "\\\\", display: "| \\\\" } \
> 	], \
> 	[ \
> 		CTRL, \
> 		ALT, \
> 		KEYBOARD, \
> 		' ', \
> 		' ', \
> 		' ', \
> 		' ', \
> 		' ', \
> 		' ', \
> 		{ key: LEFT, popup: HOME }, \
> 		{ key: DOWN, popup: PGDN }, \
> 		{ key: RIGHT, popup: END } \
> 	] \
> ]
> ```

Man u are awesome plz update this in termux wiki also mention the way to disable softkeyboard from termux app settings forever. like settings terminal I/O disable softkeyboard and done u are really life saver man

**Reply by Yonle** [reply]
I guess termux wiki does not need this. Hence it's not suggested for fast-typer.

**Reply by ghost** [reply]
> I guess termux wiki does not need this. Hence it's not suggested for fast-typer.

trust me man it's very much suggested for a fast typer just need some practice.

**Reply by SDRausty** [reply]
>suggested for fast-typer.

>very much suggested for a fast typer just need some practice.

*suggested for fast typists who just want some practice.

Possibly [brentn/zzStenoKeyboard ](https://github.com/BuildAPKs/buildAPKs.keyboards/blob/main/ma.bash#L23) can aid in stenographic typing which builds on device in Termux.  I am not a stenographer at present;  So I do not really know.  Maybe I will become a good stenographer thanks to this Android keyboard;  It builds and appears to work correctly in Termux on device when I had last tested it in Termux:
![Screenshot_20210830-150144](https://user-images.githubusercontent.com/27742457/131391035-384880f6-aa54-4d25-b8a1-2ca524c0a6d1.png)

Please reference [zzStenoKeyboard](https://github.com/brentn/zzStenoKeyboard) for published open source code of an Android stenographer keyboard that should build and install on device with Termux app.

Also please reference ["Add overlay keyboard with Alt, Ctrl, Tab, arrow keys (and a few more keys) #97"](https://github.com/termux/termux-app/issues/12) for specific Termux app `*.java` files too, and an origin of this topic here.

**Comment by agnostic-apollo** [comment]
The user can add `extra-keys-text-all-cap=false` entry to `termux.properties` file to disable auto capitalization of extra keys text for both normal and popup buttons. The default value is `true`. Running `termux-reload-settings` command will also update the behaviour instantaneously if changed. 5c72c3ca

This feature is currently only available for github debug builds, and will be available in next release `v0.118`. Should be useful for lowercasing the letters of the "alphanumeric" extra keys keyboard posted above.
